serialization.go

```diff
@@ -114,3 +116,85 @@ func CreateJSONMerkleTreeLeaf(data interface{}, timestamp uint64) *MerkleTreeLea
 		},
 	}
 }
+
+// MerkleTreeLeafFromRawChain generates a MerkleTreeLeaf from a chain and timestamp.
+func MerkleTreeLeafFromRawChain(rawChain []ASN1Cert, isPrecert bool, timestamp uint64) (*MerkleTreeLeaf, error) {
+	// Need at most 3 of the chain
+	count := 3
+	if count > len(rawChain) {
+		count = len(rawChain)
+	}
+	chain := make([]*x509.Certificate, count)
+	for i := 0; i < count; i++ {
+		cert, err := x509.ParseCertificate(rawChain[i].Data)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse chain[%d] cert: %v", i, err)
+		}
+		chain[i] = cert
+	}
+	return MerkleTreeLeafFromChain(chain, isPrecert, timestamp)
+}
+
+// MerkleTreeLeafFromChain generates a MerkleTreeLeaf from a chain and timestamp.
+func MerkleTreeLeafFromChain(chain []*x509.Certificate, isPrecert bool, timestamp uint64) (*MerkleTreeLeaf, error) {
+	leaf := MerkleTreeLeaf{
+		Version:  V1,
+		LeafType: TimestampedEntryLeafType,
+		TimestampedEntry: &TimestampedEntry{
+			Timestamp: timestamp,
+		},
+	}
+	if !isPrecert {
+		leaf.TimestampedEntry.EntryType = X509LogEntryType
+		leaf.TimestampedEntry.X509Entry = &ASN1Cert{Data: chain[0].Raw}
+		return &leaf, nil
+	}
+
+	// Pre-certs are more complicated. First, parse the leaf pre-cert and its
+	// putative issuer.
+	if len(chain) < 2 {
+		return nil, fmt.Errorf("no issuer cert available for precert leaf building")
+	}
+	issuer := chain[1]
+	cert := chain[0]
+
+	if IsPreIssuer(issuer) {
+		// The issuer of the pre-cert is not going to be the issuer of the final
+		// cert.  Change to use the final issuer.
+		if len(chain) < 3 {
+			return nil, fmt.Errorf("no issuer cert available for pre-issuer")
+		}
+		issuer = chain[2]
+
+		// Replace the cert's Issuer field with the Subject from the intermediate that will sign the final cert.
+		cert.Issuer = issuer.Subject
+		if err := cert.RebuildRawTBSCertificate(); err != nil {
+			return nil, fmt.Errorf("failed to rebuild TBS for pre-cert: %v", err)
+		}
+	}
+
+	// Next, post-process the DER-encoded TBSCertificate, to remove the CT poison
+	// extension.
+	defangedTBS, err := x509.RemoveCTPoison(cert.RawTBSCertificate)
```

Comment:
No but yes: it's important to work with the DER data of the TBS to avoid any chance of re-ordering extensions (which happens on any trip through `x509.CreateCertificate`); however, parsing from/to `tbsCertificate` is internal to the `x509` package, so is not easy to separate.

However, rather than trying to make a potentially-generally-useful `RebuildRawTBSCertificate` entrypoint, I can just generalize the CT-specific `RemoveCTPoison` entrypoint to also allow issuer change -- which means there's only one call to `asn1.Marshal` in the code path.